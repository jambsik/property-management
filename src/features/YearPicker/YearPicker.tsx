import React, { useState } from 'react';
import Input from '../Input/Input';
import { Autocomplete, TextFieldProps } from '@mui/material';

interface YearPickerProps {
    label: string;
}

const YearPicker = ({ label }: YearPickerProps) => {
    const FIRST_YEAR = new Date('01/01/1900').getFullYear();
    const LAST_YEAR = new Date().getFullYear();
    const generateYears = (firstYear: number, lastYear: number) => {
        const listOfYears = [];

        for (let year = firstYear; year <= lastYear; year++) {
            listOfYears.push({
                label: `${year}`,
            });
        }

        return listOfYears;
    };

    const [options] = useState(() => generateYears(FIRST_YEAR, LAST_YEAR));

    return (
        <Autocomplete disablePortal id="combo-box-demo" options={options} sx={{ width: '120px' }} renderInput={(params: TextFieldProps) => <Input {...params} label={label} />} />
    );
};
export default YearPicker;
