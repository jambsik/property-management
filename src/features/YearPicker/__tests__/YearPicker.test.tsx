import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import YearPicker from '../YearPicker';
describe('YearPicker component test', () => {
    type ElementFinder = (selector: string) => Element | null;

    const yearPickerLabel = 'My year picker';
    const onChangeMock = jest.fn();
    const someYearsFromNow = ['2021', '1990', '1901'];
    const inputTestOptions = {
        selector: 'li',
    };

    const pageObject = {
        input: `input[id="year-picker-${yearPickerLabel}"]`,
    };

    const typeYear = async (dispatchElementFinder: ElementFinder, element: Element, year: string, inTheDocument: boolean = true): Promise<Element | undefined> => {
        fireEvent.change(element, {
            target: {
                value: year,
            },
        });
        let option;

        await waitFor(() => {
            option = dispatchElementFinder(year);
            inTheDocument ? expect(option).toBeInTheDocument() : expect(option).not.toBeInTheDocument();
        });

        return option;
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('Should render component without errors', () => {
        const { container } = render(<YearPicker label={yearPickerLabel} onChange={onChangeMock} />);

        expect(container).toMatchSnapshot();
    });

    it('Should be able to select this or previous years', async () => {
        const { container, queryByText } = render(<YearPicker label={yearPickerLabel} onChange={onChangeMock} />);

        const inputElement = container.querySelector(pageObject.input);

        expect(inputElement).toBeInTheDocument();

        const opt1: Element | undefined = await typeYear((selector: string) => queryByText(selector, inputTestOptions), inputElement!, someYearsFromNow[0]);

        opt1 && fireEvent.click(opt1);
        expect(onChangeMock).toHaveBeenCalledTimes(1);

        const opt2: Element | undefined = await typeYear((selector: string) => queryByText(selector, inputTestOptions), inputElement!, someYearsFromNow[1]);

        opt2 && fireEvent.click(opt2);
        expect(onChangeMock).toHaveBeenCalledTimes(2);

        const opt3: Element | undefined = await typeYear((selector: string) => queryByText(selector, inputTestOptions), inputElement!, someYearsFromNow[2]);

        opt3 && fireEvent.click(opt3);
        expect(onChangeMock).toHaveBeenCalledTimes(3);
    });

    it('Should not show any option if the year is lower than 1900 or higher than the current year', async () => {
        const nextYear = `{new Date().getFullYear() + 1}`;

        const { container, queryByText } = render(<YearPicker label={yearPickerLabel} onChange={onChangeMock} />);

        const inputElement = container.querySelector(pageObject.input);

        expect(inputElement).toBeInTheDocument();

        const opt1: Element | undefined = await typeYear((selector: string) => queryByText(selector, inputTestOptions), inputElement!, '1800', false);

        opt1 && fireEvent.click(opt1);
        expect(onChangeMock).toHaveBeenCalledTimes(0);

        const opt2: Element | undefined = await typeYear((selector: string) => queryByText(selector, inputTestOptions), inputElement!, nextYear, false);

        opt2 && fireEvent.click(opt2);
        expect(onChangeMock).toHaveBeenCalledTimes(0);
    });

    it('Should dispatch on change after select an option', async () => {
        const { container, queryByText } = render(<YearPicker label={yearPickerLabel} onChange={onChangeMock} />);

        const inputElement = container.querySelector(pageObject.input);

        expect(inputElement).toBeInTheDocument();

        const option: Element | undefined = await typeYear((selector: string) => queryByText(selector, inputTestOptions), inputElement!, someYearsFromNow[0]);

        option && fireEvent.click(option);
        expect(onChangeMock).toHaveBeenCalledTimes(1);
    });
});
