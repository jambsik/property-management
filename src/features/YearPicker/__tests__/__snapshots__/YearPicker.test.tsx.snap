// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearPicker component test Should render component without errors 1`] = `
<div>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-mmax81-MuiAutocomplete-root"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-jemhih-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorInfo css-ahcq22-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="year-picker-My year picker"
        id="year-picker-My year picker-label"
      >
        My year picker
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorInfo MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-194ngh9-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id="year-picker-My year picker"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
